pandas/core/categorical.py

```diff
@@ -1364,6 +1364,29 @@ def sort_values(self, inplace=False, ascending=True, na_position='last'):
             return self._constructor(values=codes, categories=self.categories,
                                      ordered=self.ordered, fastpath=True)
 
+    def _rank(self, *args, **kwargs):
+        """
+        For correctly ranking ordered categorical data. See GH#15420
+
+        Ordered categorical data should be ranked on the basis of
+        codes.
+
+        Returns
+        -------
+        numpy array
+
+        """
+        from pandas.core.algorithms import rank
+
+        if self._ordered:
+            codes = self._codes.astype('float64')
+            na_mask = (codes == -1)
+            codes[na_mask] = np.nan
```

Comment:
you are uncessarily converting to object and we are not longer calling rank directly here. (and you don't need to pass thru args, kwargs).
```
if self.ordered:
    values = self.codes
    mask = values == -1
    if mask.any():
        values = values.astype('float64')
        values[mask] = np.nan
else:
    values = np.array(values)
```